### PR TITLE
3rdparty: update libpng 1.6.43

### PR DESCRIPTION
**Pull request by sturkmen72**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
@sturkmen72 Thanks a lot for the update. Brief review summary:
- 20190528-fix-leak-png_handle_exif.diff already integrated to libpng and not needed.
- 20190910-msa-patch.diff - looks like too.
- The patch contains code for Loongarch and some new optimizations for MIPS, but they are not handled in CMakeLists.txt
- `#if defined(_MSC_VER) && !defined(__clang__) && defined(_M_ARM64)` looks suspicious. It's present in original code, but it affects performance on Mac and Android. I'll check it and provide more feedback.

**Comment by sturkmen72** [comment]
@asmorkalov thanks for review.

i removed the patch files.
i will look CMakeLists.txt to handle new optimizations ( any help is welcome )

**Comment by asmorkalov** [comment]
@fengyuentau @WanliZhong Could you help with Loongarch?

**Comment by fengyuentau** [comment]
> @fengyuentau @WanliZhong Could you help with Loongarch?

Do you mean changes in CMake to support new optimizations for loongarch?

**Comment by asmorkalov** [comment]
Yes, test locally and ensure that optimizations are enabled correctly.

**Comment by fengyuentau** [comment]
```
$ cmake -B build -DBUILD_PNG=ON .
...
--   Media I/O:
--     ZLib:                        /usr/lib/loongarch64-linux-gnu/libz.so (ver 1.2.11)
--     JPEG:                        build-libjpeg-turbo (ver 2.1.3-62)
--       SIMD Support Request:      YES
--       SIMD Support:              NO
--     WEBP:                        build (ver encoder: 0x020f)
--     PNG:                         build (ver 1.6.43)
--     TIFF:                        build (ver 42 - 4.6.0)
--     JPEG 2000:                   build (ver 2.5.0)
--     OpenEXR:                     build (ver 2.3.0)
--     HDR:                         YES
--     SUNRASTER:                   YES
--     PXM:                         YES
--     PFM:                         YES
...
```

@asmorkalov  How do I know wheter

> ensure that optimizations are enabled correctly

**Comment by sturkmen72** [comment]
@fengyuentau thank you for your help and patience.  after applying my last commit could you post please all cmake output.

**Comment by sturkmen72** [comment]
@asmorkalov @fengyuentau for better testing is it possible temporarily to use https://github.com/opencv/ci-gha-workflow/blob/main/.github/workflows/OCV-PR-4.x-loongnix-loongarch64.yaml

**Comment by fengyuentau** [comment]
I got this error

```
-- Found assembler: /usr/bin/cc
CMake Error at 3rdparty/libpng/CMakeLists.txt:185 (CHECK_C_COMPILER_FLAG):
  Unknown CMake command "CHECK_C_COMPILER_FLAG".


-- Configuring incomplete, errors occurred!
```

**Comment by sturkmen72** [comment]
> I got this error
> 
> ```
> -- Found assembler: /usr/bin/cc
> CMake Error at 3rdparty/libpng/CMakeLists.txt:185 (CHECK_C_COMPILER_FLAG):
>   Unknown CMake command "CHECK_C_COMPILER_FLAG".
> 
> 
> -- Configuring incomplete, errors occurred!
> ```

my bad.  i missed  include(CheckCCompilerFlag)
.
.
if(TARGET_ARCH MATCHES "^(loongarch)")
  include(CheckCCompilerFlag)
  set(PNG_LOONGARCH_LSX_POSSIBLE_VALUES on off)
.
.

**Comment by sturkmen72** [comment]
@fengyuentau thank you very much for your help. @asmorkalov let me try to add PNG optimization related information in cmake output (to be seen in getBuildInformation())

**Comment by asmorkalov** [comment]
Looks like we have troubles with ninja-build:
```
-- Configuring done
CMake Error: Error required internal CMake variable not set, cmake may not be built correctly.
Missing variable is:
CMAKE_ASM_COMPILE_OBJECT
-- Generating done
CMake Error:
  Running

   '/opt/android-sdk/cmake/3.22.1/bin/ninja' '-C' '/home/ci/build/o4a' '-t' 'recompact'

  failed with:

   ninja: error: build.ninja:35: loading 'CMakeFiles/rules.ninja': No such file or directory

  include CMakeFiles/rules.ninja

                                ^ near here
```

**Comment by sturkmen72** [comment]
in my last failed commit it was `  project(${PNG_LIBRARY} C ASM)`

and now i reverted the change.. lets see if this solves build problem of android.
```
if(ENABLE_NEON)
  project(${PNG_LIBRARY} C ASM)
else()
  project(${PNG_LIBRARY} C)
endif()
```

**Review by asmorkalov — COMMENTED** [review]
Thanks a lot for the contribution! The patch looks good in general. I just fixed a typo.
I propose to modify CMake output a bit and make it similar to libjpeg-turbo.

**Comment by sturkmen72** [comment]
> Thanks a lot for the contribution! The patch looks good in general. I just fixed a typo. I propose to modify CMake output a bit and make it similar to libjpeg-turbo.

you mean
```
PNG:                         build (ver 1.6.43)
  Hardware Optimizations:    YES
```
or
```
PNG:                         build (ver 1.6.43)
  SIMD:                      YES
```

**Comment by asmorkalov** [comment]
Something like:
``
--       SIMD Support Request:      YES
--       SIMD Support:              YES (Arm NEON)
```

**Comment by sturkmen72** [comment]
@asmorkalov @fengyuentau thank you for your help. i think the PR is almost ready to merge. there is still one small issue in my opinion. let me have some time to solve it.

**Comment by asmorkalov** [comment]
Made brief performance testing. No changes since previous version. Also build with zlib-ng brings performance gain:
11th Gen Intel(R) Core(TM) i5-1135G7 @ 2.40GHz
```
Name of Test new-png-1 new-png-zlib-ng-1 new-png-zlib-ng-1
                                                vs        
                                             new-png-1    
                                            (x-factor)    
Decode::JPEG  33.928        34.182             0.99       
Encode::JPEG  31.841        31.958             1.00       
decode::PNG   23.355        12.801             1.82       
encode::PNG   49.293        30.316             1.63 
```

File: 3rdparty/libpng/CMakeLists.txt [hunk 1]
```diff
@@ -9,63 +9,230 @@ else()
...
+# enable. Note that this will fail if you attempt to build a universal binary in
+# a single CMake invocation.
+if (APPLE AND CMAKE_OSX_ARCHITECTURES)
+  set(TARGET_ARCH ${CMAKE_OSX_ARCHITECTURES})
+else()
+  set(TARGET_ARCH ${CMAKE_SYSTEM_PROCESSOR})
+endif()
+
+OCV_OPTION(PNG_HARDWARE_OPTIMIZATIONS "Enable Hardware Optimizations, if available for this platform" (NOT CV_DISABLE_OPTIMIZATION)
+  VISIBLE_IF BUILD_PNG)
```
Comment (asmorkalov) [inline 3rdparty/libpng/CMakeLists.txt:62]: Duplicate too.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1473,6 +1473,31 @@ if(WITH_SPNG)
...
+	  elseif(PNG_MIPS_MSA AND NOT PNG_MIPS_MMI)
+	    status("      SIMD Support:" "YES (Mips MSA)")
+	  else()
+	    status("      SIMD Support:" "YES (Mips MMI)")
+	  endif()
+	elseif(PNG_LOONGARCH_LSX)
+	  status("      SIMD Support:" "YES (LoongArch LSX)")
+	else()
+	  status("      SIMD Support:" "NO")
+	endif()
```
Comment (asmorkalov) [inline CMakeLists.txt:1496]: Formatting.

**Comment by sturkmen72** [comment]
> @asmorkalov @fengyuentau thank you for your help. i think the PR is almost ready to merge. there is still one small issue in my opinion. let me have some time to solve it.

I noticed that the "cmake -DCV_DISABLE_OPTIMIZATION=ON" command worked but changing CV_DISABLE_OPTIMIZATION value in cmake-gui did nothing. I looked for a solution to this but couldn't find it. can be looked at later, beyond the scope of this PR.

i think the PR is ready to merge.

**Review by asmorkalov — APPROVED** [review]

**Pull request merged**